core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -156,14 +188,25 @@ public void setCompression(String compressorName) {
 
   @Override
   public boolean isReady() {
+    if (internalClosed) {
+      return false;
+    }
     return stream.isReady();
   }
 
   @Override
   public void close(Status status, Metadata trailers) {
     checkState(!closeCalled, "call already closed");
     closeCalled = true;
-    stream.close(status, trailers);
+    if (internalClosed) {
+      return;
+    }
+
+    if (status.isOk() && method.getType().serverSendsOneMessage() && !messageSent) {
+      internalClose(Status.INTERNAL.withDescription(MISSING_RESPONSE));
+    } else {
+      stream.close(status, trailers);
```

Comment:
Prefer making error cases return to avoid indentation in the common case. Sometimes it doesn't matter, but here it seems like it would improve the readability.